<a href="https://colab.research.google.com/github/vvtrip/ml_manifestations/blob/master/stats/4_stats_with_python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
from scipy import stats
import statistics

# Basics

**Descriptive Statistics** 
 - it summaries a dataset, which helps in gaining insights, and making inferences about a dataset.
 - In general, we compute statistical measures of one or more samples, related to a population, and draw conclusions about the population.
 - Descriptive Statistics involves estimating centrality measures and measures of dispersion.

**Centrality Measures**
 - It determine the center of a dataset.
 - The three major centrality measures are: mean, median, and mode.

**Mean**
 - it is the sum of all values divided by a total number of values, of a data set.
 - mean function of numpy is used to compute mean of an array of numbers

**Median**
 - it is the value that separates the given data set into two halves.
 - median function of numpy library can be used to compute median of a data set.

**Mode**
 - it is the value that appears most often in a dataset is called the mode value.
 - mode function of scipy.stats module can be used for computing mode of a given data set.

**Measures of Dispersion**
 - it provides insights on the spread of given dataset.
 - Major measures of dispersion are: range, percentile, inter-quartile range, standard deviation, variance, skewness, and kurtosis.


**Range**
 - it is the difference between maximum and minimum values of the dataset.
- import numpy as np
- s1 = np.array([86, 47, 45, 47, 40])
- print(np.ptp(s1))

**Percentile**
 - it refers to a value, below which lies given the percentage of data points.

 - E.g., 45th percentile refers to a value below which 45% of data points are found.
 - percentile function of numpy can be used to compute a single or multiple percentiles.

**Quartiles**
 - Three Quartiles namely, Q1 Q2 and Q3, split the entire dataset into four equal parts.
 - Each part contains 25% of data.

**Inter Quartile Range (IQR)**
 - it refers to difference between third quartile (Q3) and first quartile (Q1).
 - iqr method from scipy.stats can be used for calculating it.

**Variance**
 - it is defined as the average of squared differences, of each data point from dataset's mean.

**Standard Deviation**
 - it is square root of variance.
 - var and std functions of numpy can be used for computing variance and standard deviation respectively.
 - By default, the functions assume that the dataset represents entire population.
-  To represent a sample, derived from a population, ddof parameter(delta degree of freedom is set to 1.

**Skewness**
- it determines whether the majority of data points are present on one side of the distribution.
 - A positive value represents right skewed distribution; a negative value represents left skewed one, zero represent unskewed distribution.

**Kurtosis**
 - Kurtosis indicates how much of data is concentrated around mean or shape of the probability distribution.
 - It can be estimated using kurtosis function of scipy.stats module.
 - By default, it uses Fisher’s definition. This can be changed to Pearson by setting fisher parameter to False.

In [2]:
s = [26, 15, 8, 44, 26, 13, 38, 24, 17, 29]
print(f"Mean is {np.mean(s)}")
print(f"Meadian is {np.median(s)}")
print(f"Mode is {stats.mode(s)}, {stats.mode(s)[0][0]}")
print(f"without interpolation 1st and 3rd quartiles are {np.percentile(s,[25,75])}")
print(f"without interpolation IQR through scipy.stats is {stats.iqr(s, rng=(25,75))}")
print(f"with interpolation 1st and 3rd quartiles are {np.percentile(s,[25,75], interpolation='lower')}")
print(f"with interpolation IQR through scipy.stats is {stats.iqr(s, rng=(25,75), interpolation='lower')}")
print(f"Skewness is {round(stats.skew(s),2)}")
print(f"Kurtosis is {round(stats.kurtosis(s),2)}")

Mean is 24.0
Meadian is 25.0
Mode is ModeResult(mode=array([26]), count=array([2])), 26
without interpolation 1st and 3rd quartiles are [15.5  28.25]
without interpolation IQR through scipy.stats is 12.75
with interpolation 1st and 3rd quartiles are [15 26]
with interpolation IQR through scipy.stats is 11
Skewness is 0.36
Kurtosis is -0.77


# Random Variables, Random Numbers and Random Distributions

A random number is a number, chosen by chance from a distribution.
Python provides a lot of modules, which deal with random numbers.
- random module of python standard library.
- random module of Numpy
- stats module of Scipy

random module of numpy has utilities, which generate arrays of random numbers.
- E.g.: rand function generates uniformly distributed numbers from range [0, 1]

- rand function with no arguments generate a single random value.

- By passing arguments, it generates a random array of specified size.

In [3]:
import numpy as np
print(np.random.rand())
# generates a 2*3 array
print(np.random.rand(2,3))

0.8431032824112108
[[0.47596183 0.56323604 0.6357884 ]
 [0.90152453 0.12532934 0.45534898]]


- In statistics, you select items randomly from a population, either with or without a replacement.
- This can be achieved with choice method

In [4]:
import numpy as np
print(np.random.choice([11, 22, 33], 2, replace=False))

[33 22]


**Random Seeding**
- Seed is an important concept when it comes to reproducibility. If you are working with random numbers and you would want to peers to validate your results, i.e., they should also get the same random sequence as you did, you can set the seed to a particular value and send the seed value to your peers.
- seed is a number that sets the initial state of random number generator.
- Setting a seed, helps in generating the same sequence of random numbers, repeatedly.
- seed method of a random module can be used to set a seed as shown in below example.

In [5]:
import numpy as np

np.random.seed(100)
print(np.random.rand())

np.random.seed(100)
print(np.random.rand())

0.5434049417909654
0.5434049417909654


**Random Variables**
 - In probability theory, the set of all possible outcomes of a random experiment is known as sample space.
 - Probabilities of all outcomes of the experiment define the probability distribution.
 - A random variable is a variable that takes real numbers or integers and map each value to one of the outcomes of sample space. 
 - E.g.: In an experiment of tossing a coin, the sample space is {'Head', 'Tail'} and a possible random variable takes the value 0 for head and 1 for the tail.

**Probability Distributions**
 - There are two types of probability distributions namely discrete and continuous that take integer and real values, respectively.
 - scipy.stats module provides classes that represent random variables, corresponding to a large number of probability distributions.
 - E.g: the class norm represent normal continuous random variable, and binom represent binomial discrete random variable.

- scipy.stats module provide a lot of methods for created discrete and continuous random variables.
- Commonly used methods are :
- pdf / pmf : Probability distribution function (continuous) or probability mass function (discrete).
- cdf : Cumulative distribution function.
- sf : Survival function (1 – cdf).
- rvs : Creating random samples from a distribution.
- The following example defines a normal continuous random variable of mean 1.0 and std 2.5.
- It also estimates probabilities and cumulative probabilities at -1, 0 and 1.
- The example also generates six random numbers from defined normal distribution.

In [6]:
from scipy import stats

x = stats.norm(loc=1.0, scale=2.5)

print(x.pdf([-1, 0, 1]))

print(x.cdf([-1, 0, 1]))

print(x.rvs((2,3)))

[0.11587662 0.14730806 0.15957691]
[0.2118554  0.34457826 0.5       ]
[[-0.40409689 -3.12269376  1.88668613]
 [-0.96516083  0.42031951  1.5199392 ]]


In [7]:
# create a normal distribution with mean 32 and standard deviation 4.5
# set the random seed to 1, and create a random sample of 100 elements from the above distribution
# compute aboslute difference between the sample mean and distribution mean

import numpy as np
from scipy import stats
np.random.seed(1)

x = stats.norm.rvs(loc=32, scale=4.5, size=100)
print(round(abs(np.mean(x)-32),2))

0.27


In [8]:
# Simulate a random experiment of tossing a coin 10000 times and determine the count of Heads.
# Hint: Define a binomial distribution with n = 1 and p = 0.5

import numpy as np
from scipy import stats
np.random.seed(1)

x = stats.binom(1, 0.5)
y = x.rvs(10000)
print(np.bincount(y)[0])

4990


# Hypothesis testing with scipy


- Hypothesis Testing is a methodology for evaluating if a claim is acceptable or not, based on data.
- In a Hypothesis Testing, a Null Hypothesis (Ho) represents currently accepted the state of knowledge, and an Alternative Hypothesis (Ha) represents a new claim which challenges the currently accepted state of knowledge.
- The null hypothesis and the alternative hypothesis are mutually exclusive.

**Steps Involved in Hypothesis Testing**
- The following steps are involved in a Hypothesis Testing:
 - Define the null hypothesis and the alternative hypothesis.
 - Select a test statistics whose probability distribution function can be found under the null hypothesis.
 - Collect data.
 - Compute the test statistics from the data and calculate its p-value under the null hypothesis.
 - Null hypothesis is rejected if the p-value is lower than predetermined significance value.

**Choosing Test Statistics**
 - In hypothesis testing, selecting a test statistics is the most difficult part.
 - The methods used for performing t-test are shown below.
 - stats.ttest_1samp: Tests if the mean of a population is a given value.
 - stats.ttest_ind: Tests if the means of two independent samples are equal.
 - stats.ttest_rel: Tests if the means of two paired samples are equal.

**Example 1**
 - Let's consider a common hypothesis: Mean of a population is equal to certain value.
 - In reality, we estimate mean and variance of a sample and calculate the test statistic.
 - If population variance is identified, then it is reasonable to consider that test statistic is normally distributed.
 - If population variance is unknown, sample variance is used, and test statistic follows t distribution.

In [9]:
# Consider a normal population with mean 0.8 and standard deviation 0.5.
# Define the null hypothesis as Mean of the population is 1.0.
# Let's calculate t-statistic and p-value

from scipy import stats
import numpy as np
np.random.seed(1)

mu, sigma = 0.8, 0.5
X = stats.norm(mu, sigma)

# Deriving a sample
n = 100
X_sample = X.rvs(n)

# Computing test statistic
t, p = stats.ttest_1samp(X_sample, 1.0)
print(round(t,2), p)

-3.82 0.00023686273495632666


In previous example, the obtained t-statistic value is 3.82 and p-value is 0.0000023686....

Since p-value is very low and less than the significance level 0.05, you can reject the null hypothesis and infer that the mean of the population is not 1.

In [10]:
# Let's consider another problem, where the null hypothesis states 
# that the population means of two random variables are equal.

# The below example derives two samples from different populations 
# and verifies the claim that their population means are equal.

X1 = stats.norm(0.25, 1.0)
X2 = stats.norm(0.50, 1.0)
np.random.seed(1)

X1_sample = X1.rvs(100)
X2_sample = X2.rvs(100)

t, p = stats.ttest_ind(X1_sample, X2_sample)
print(round(t,2), p)

-2.65 0.008722880782836376


In previous example, the obtained t-statistic value is -2.65 and p-value is 0.008722...

Since p-value is very low and less than the significance level 0.05, you can reject the null hypothesis and state that population means of both samples differ.

In [11]:
import numpy as np
from scipy import stats

# The following are samples represent life satisfaction score(through a methodology)
# of older adults and younger adults respectively. Compute t-statistics amnd display the
# t-score and p-value 
# sol - these are independent samples

s1 = [45, 38, 52, 48, 25, 39, 51, 46, 55, 46]
s2 = [34, 22, 15, 27, 37, 41, 24, 19, 26, 36]

t, p = stats.ttest_ind(s1, s2)

print(t, p)

4.257546665558161 0.0004736633119019225


In [12]:
#Write your code here

import numpy as np
from scipy import stats

# following are samples of chocolate chip consumption by 10 rats with and 
#  without electrical simulation. calculate t-statistics and provide t-score and p-value
# sol - these are related samples

s1 = [12, 7, 3, 11, 8, 5, 14, 7, 9, 10]
s2 = [8, 7, 4, 14, 6, 7, 12, 5, 5, 8]

t, p = stats.ttest_rel(s1, s2)

print(t, p)

1.315587028960544 0.2208380130273219


# Statistical Modelling

 - A Statistical Model is a mathematical equation, which explains the relationship between dependent variables (Y) and independent variables (X).
 - In general, a model is written as Y = f(X)
However, in reality, an element of uncertainty is expected due to factors such as measurement noise. Hence the aforementioned equation can be rewritten as Y = f(X) + e   'e' is residual error

**Statistical Modelling**
- Statistical Modelling deals with creating models that attempt to explain the data best.
- The simplest model is a linear model represented as Y = B0 + B1*X + e, where the coefficients, B0 and B1 are the parameters of the model and e is normally distributed residual error.
- A linear regression model assumes that residuals are independent and normally distributed.
- The model is fitted to data using ordinary least squares approach.

**Linear Models**
 - In most of the linear regression models, a dependent variable y is written as :
 - a linear combination of the response variables X i.e y = B0 + B1*X1 + ... + Bn*Xn, or
functions of the response variables i.e y = B0 + B1*X + B2*x^2 + ... + Bn*X^n, or
 - models that have a linear component i.e y = B0 + B1*sin(X1) + B2*cos(X2)

**Non-Linear Models**
 - Other than linear regression models, statistical modeling can be used to build non-linear models.

 - Errors of dependent variable follow a distribution other than normal distribution.

 - Examples of non-linear models are: Binomial Regression and Poisson Regression.

 - In most of the cases, the non-linear models are generalized to linear models.

**Design Matrices**
 - In reality, you choose a model and fit the available data to it.

 - Once a model is chosen, design matrices y and X are constructed, and the regression problem is written, in matrix form, as y = XB + e.

 - where y is the vector of dependent variables, X is the vector of independent variables, B is a vector of coefficients, and e is the residual (error).

 - Thus obtained design matrices are passed as inputs to the chosen model.

**Statistical Modelling with StatsModels**
 - statsmodels library supports several types of statistical models.

 - All of them follow a similar usage pattern.

 - A Statistical model is represented by a model class.

 - A model can be initiated with,

  - given design matrices of dependent and independent variables, or with
  - given patsy formula and data frame or dictionary-like object.

**Step 1: Creating a Model**
 - An instance of a model class is created in either of the following ways.

 - model = sm.MODEL(y, X) or
 - model = smf.model(patsy_formula, data)
 
 where MODEL and model refer to model names such as OLS, GLM, ols, glm, etc.

 Uppercase names take design matrices as arguments, and lowercase names take Patsy formulas and data frames as arguments.

**Step 2: Fitting a Model**

In order to fit the model with data, fit method is invoked on the created model, as shown below.

 - result = model.fit()

The fit method returns a result object, which has methods and attributes for further analysis.


**Step 3: Viewing Model Summary**
 - The summary method of result object produces a summary text that describes the result of the fit, as shown below.

 - print(result.summary())

 - The displayed summary text varies for each statistical model and provides information of various statistical parameters.

**Step 4: Analyzing the Model Further**
Other than viewing summary statistics, you can perform activities like,

 - Determining fitted values,
 - Predicting the dependent variable values for new independent variable values.
 - Checking if residuals of fitted models follow a normal distribution or not.

# Statistical modelling with Patsy

**Constructing Design Matrices**

The below example shows calculation of design matrices, y, and X, for the considered linear model Y = B0 + B1*X1 + B2*X2 + B3*X1*X2.

In [13]:
import numpy as np

y = np.array([1, 2, 3, 4, 5])
x1 = np.array([6, 7, 8, 9, 10])
x2 = np.array([11, 12, 13, 14, 15])
X = np.vstack([np.ones(5), x1, x2, x1*x2]).T

print(y)
print(X)

# Thus obtained design matrices (y and X) can be passed to regression methods for obtaining the coefficient vector.

[1 2 3 4 5]
[[  1.   6.  11.  66.]
 [  1.   7.  12.  84.]
 [  1.   8.  13. 104.]
 [  1.   9.  14. 126.]
 [  1.  10.  15. 150.]]


**Design Matrices with patsy**
 - patsy, a  python library, allows defining a model in simpler easily.

 - It also constructs relevant design matrices, automatically, using patsy.dmatrices function.

 - patsy.dmatrices takes a formula (in string form) as a first argument, and a dictionary-like object with data arrays for the response variables as second arguments.

In [14]:
import patsy
import numpy as np

y = np.array([1, 2, 3, 4, 5])
x1 = np.array([6, 7, 8, 9, 10])
x2 = np.array([11, 12, 13, 14, 15])
data = {'y':y, 'x1':x1, 'x2':x2}

y, X = patsy.dmatrices('y ~ 1 + x1 + x2 + x1*x2', data)

print(y)
print(X)

[[1.]
 [2.]
 [3.]
 [4.]
 [5.]]
[[  1.   6.  11.  66.]
 [  1.   7.  12.  84.]
 [  1.   8.  13. 104.]
 [  1.   9.  14. 126.]
 [  1.  10.  15. 150.]]


**Understanding patsy Formulae**

Let's understand few patsy formulae provided below.

 - 'y ~ x' : y is linearly dependent on x. ~ symbol separates dependent variable from independent variable terms. It is also equivalent to 'y ~ 1 + x'.

 - 'y ~ x1 + x2' : y is a linear combination of x1 and x2. + sign is used to denote the union of terms.

 - y ~ x1*x2 : x1*x2 is an interaction term that includes all lower order terms. Hence formula is equivalent to y ~ 1 + x1 + x2 + x1*x2.

  - 'y ~ np.log(x1)': Often numpy functions can be used to transform terms in the expression.

 - 'y ~ I(x1 + x2)': I is the identify function, used to escape arithmetic expressions and are evaluated.

 - 'y ~ C(x1)': Treats the variable x1 as a categorical variable.

**Example Datasets**
 - statsmodels contain few popular example datasets, which can be used to explore various utilities of the package.

 - The example data sets are available in datasets module.

 - Each dataset is associated with special variables like SOURCE, DESCSHORT, DESCLONG, that provide more info about the dataset.

 - A dataset can be loaded using load function, and its data can be accessed using data attribute in the form of Numpy's recarray.

**Loading Example Datasets**
 - The below example shows loading of popular breast cancer dataset and accessing its data as numpy array.

In [15]:
import statsmodels.api as sm

bc_cancer_set = sm.datasets.cancer

bc_cancer = bc_cancer_set.load()

bc_cancer_data = bc_cancer.data

print(type(bc_cancer_data))

<class 'numpy.recarray'>


**Loading R Datasets**

statsmodels package provides access to many example datasets of R, listed at R Datasets repository (https://vincentarelbundock.github.io/Rdatasets/datasets.html)

The below example loads Icecream dataset from Ecdat package.

The data is accessible via data attribute, as a pandas data frame.

In [16]:
import statsmodels.api as sm

icecream_data = sm.datasets.get_rdataset('Icecream', 'Ecdat')
data = icecream_data.data

print(icecream_data.data.shape)

(30, 4)
